src/ui/menu_nbgl.c

```diff
@@ -0,0 +1,144 @@
+        content->infosList.infoContents = (const char**) info_contents;
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+static void settings_controls_callback(int token, uint8_t index) {
+    UNUSED(index);
+    uint8_t new_setting;
+    switch (token) {
+        case BLIND_SIGNING_TOKEN:
+            G_switches[BLIND_SIGNING_IDX].initState = !(G_switches[BLIND_SIGNING_IDX].initState);
+            new_setting = (G_switches[BLIND_SIGNING_IDX].initState == ON_STATE);
+            nvm_write((void*) &N_storage.settings.allow_blind_sign,
+                      &new_setting,
+                      sizeof(new_setting));
+            break;
+        case PUBLIC_KEY_LENGTH_TOKEN:
+            G_switches[PUBLIC_KEY_LENGTH_IDX].initState =
+                !(G_switches[PUBLIC_KEY_LENGTH_IDX].initState);
+            new_setting = (G_switches[PUBLIC_KEY_LENGTH_IDX].initState == ON_STATE);
+            nvm_write((void*) &N_storage.settings.pubkey_display,
+                      &new_setting,
+                      sizeof(new_setting));
+            break;
+        case DISPLAY_MODE_TOKEN:
+            G_switches[DISPLAY_MODE_IDX].initState = !(G_switches[DISPLAY_MODE_IDX].initState);
+            new_setting = (G_switches[DISPLAY_MODE_IDX].initState == ON_STATE);
+            nvm_write((void*) &N_storage.settings.display_mode, &new_setting, sizeof(new_setting));
+            break;
+        default:
+            PRINTF("Unreachable\n");
+            break;
+    }
+}
+
+static void ui_menu_settings(void) {
+    G_switches[BLIND_SIGNING_IDX].text = "Blind signing";
+    G_switches[BLIND_SIGNING_IDX].subText = "Enable blind signing";
+    G_switches[BLIND_SIGNING_IDX].token = BLIND_SIGNING_TOKEN;
+    G_switches[BLIND_SIGNING_IDX].tuneId = TUNE_TAP_CASUAL;
+    if (N_storage.settings.allow_blind_sign == BlindSignDisabled) {
+        G_switches[BLIND_SIGNING_IDX].initState = OFF_STATE;
+    } else {
+        G_switches[BLIND_SIGNING_IDX].initState = ON_STATE;
+    }
+
+    G_switches[PUBLIC_KEY_LENGTH_IDX].text = "Public key length";
+    G_switches[PUBLIC_KEY_LENGTH_IDX].subText = "Display short public keys";
+    G_switches[PUBLIC_KEY_LENGTH_IDX].token = PUBLIC_KEY_LENGTH_TOKEN;
+    G_switches[PUBLIC_KEY_LENGTH_IDX].tuneId = TUNE_TAP_CASUAL;
+    if (N_storage.settings.pubkey_display == PubkeyDisplayLong) {
+        G_switches[PUBLIC_KEY_LENGTH_IDX].initState = OFF_STATE;
+    } else {
+        G_switches[PUBLIC_KEY_LENGTH_IDX].initState = ON_STATE;
+    }
+
+    G_switches[DISPLAY_MODE_IDX].text = "Display mode";
+    G_switches[DISPLAY_MODE_IDX].subText = "Use Expert display mode";
+    G_switches[DISPLAY_MODE_IDX].token = DISPLAY_MODE_TOKEN;
+    G_switches[DISPLAY_MODE_IDX].tuneId = TUNE_TAP_CASUAL;
+    if (N_storage.settings.display_mode == DisplayModeUser) {
+        G_switches[DISPLAY_MODE_IDX].initState = OFF_STATE;
+    } else {
+        G_switches[DISPLAY_MODE_IDX].initState = ON_STATE;
+    }
```

Comment:
If you put this in `settings_nav_callback()` you can simplify `settings_controls_callback()` so that it only handles NVM part but not the switch part.

Reply:
Good to know